spark3/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -64,16 +69,28 @@
     this.spark = spark;
     this.table = table;
     this.options = options;
+    this.snapshotId = Spark3Util.propertyAsLong(options, SparkReadOptions.SNAPSHOT_ID, null);
+    this.asOfTimestamp = Spark3Util.propertyAsLong(options, SparkReadOptions.AS_OF_TIMESTAMP, null);
     this.caseSensitive = Boolean.parseBoolean(spark.conf().get("spark.sql.caseSensitive"));
   }
 
+  private Schema snapshotSchema() {
+    if (snapshotId != null && table instanceof BaseTable) {
+      return ((BaseTable) table).schemaForSnapshot(snapshotId);
+    } else if (asOfTimestamp != null && table instanceof BaseTable) {
+      return ((BaseTable) table).schemaForSnapshotAsOfTime(asOfTimestamp);
+    } else {
+      return table.schema();
+    }
```

Comment:
Same as above, I think this should use the snapshot's schema or the table schema for now. Recovery can happen in a later update.